cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java

```diff
@@ -217,8 +259,80 @@ private Lineage doComputeLineage(final NamespacedEntityId sourceData, long start
       }
     }
 
-    Lineage lineage = new Lineage(
-      Iterables.concat(Maps.transformValues(relations.asMap(), COLLAPSE_UNKNOWN_TYPE_FUNCTION).values()));
+    Lineage lineage;
+    if (rollup.contains("workflow")) {
+      // Get workflow Run IDs for all the programs in the relations
+      final Set<String> workflowIDs = new HashSet<>();
+      for (RelationKey key : relations.keySet()) {
+        Collection<Relation> allRelations = relations.get(key);
+        for (Relation relation : allRelations) {
+          ProgramId program = relation.getProgram();
+          RunRecordMeta runRecord = store.getRun(program, relation.getRun().getId());
+          if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")) {
+            String workflowRunId = runRecord.getProperties().get("workflowrunid");
+            workflowIDs.add(workflowRunId);
+          }
+        }
+      }
+
+      // Get Program IDs for workflow Run IDs
+      Map<ProgramRunId, RunRecordMeta> workflowRunRecordMap =
+        store.getRuns(ProgramRunStatus.ALL,
+                      new Predicate<RunRecordMeta>() {
+                        @Override
+                        public boolean apply(RunRecordMeta input) {
+                          return workflowIDs.contains(input.getPid());
+                        }
+                      });
+
+      // Create a map from RunId to ProgramId for all workflows
+      Map<String, ProgramRunId> workflowIdMap = new HashMap<>();
+      for (Map.Entry<ProgramRunId, RunRecordMeta> entry : workflowRunRecordMap.entrySet()) {
+        workflowIdMap.put(entry.getValue().getPid(), entry.getKey());
+      }
+
+      // For all relations, replace ProgramIds with workflow ProgramIds
+      Multimap<RelationKey, Relation> relationsNew = HashMultimap.create();
+
+      for (RelationKey key : relations.keySet()) {
+        Collection<Relation> allRelations = relations.get(key);
```

Comment:
Instead of looping through keys and get the value (which results in n lookups in the map), it's better to iterate over the map entries.

```java
for (Map.Entry<RelationKey, Collection<Relation>> entry : relations.entrySet()) {
  ...
}
```